src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1513 @@
+        if attention_mask is not None:
+            extended_attention_mask = (1.0 - attention_mask[None, :, None, :]) * -10000.0
+            extended_attention_mask = extended_attention_mask.expand((self.ngram, batch_size, seq_length, seq_length))
+            extended_attention_mask = torch.cat([extended_attention_mask, extended_attention_mask], dim=-1)
+            return extended_ngram_causal_mask + extended_attention_mask
+        return extended_ngram_causal_mask
+
+    def forward(
+        self,
+        input_ids=None,
+        attention_mask=None,
+        encoder_hidden_states=None,
+        encoder_attention_mask=None,
+        past_key_values=None,
+        inputs_embeds=None,  # TODO(Patrick) Implement
+        use_cache=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+        **unused,
+    ):
+        """
+
+        Args:
+            input_ids (LongTensor): previous decoder outputs of shape
+                `(batch, tgt_len)`, for teacher forcing
+            encoder_hidden_states: output from the encoder, used for
+                encoder-side attention
+            encoder_attention_mask: for ignoring pad tokens
+            decoder_cached_states (dict or None): dictionary used for storing state during generation
+            use_cache: inference or training procedure.
+
+        Returns:
+            tuple:
+                - the decoder's features of next n-grams, with shape `(batch, self.ngram * tgt_len, embed_dim)`
+                - hidden states
+                - attentions
+
+        """
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+
+        if encoder_attention_mask is not None:
+            # invert mask
+            encoder_attention_mask = encoder_attention_mask.eq(0)
+
+        main_stream_pos_embed, real_positions = self.embed_positions(
+            input_ids, use_cache=use_cache, past_key_values=past_key_values
+        )
+
+        if use_cache:
+            i_buckets_main_stream, i_bucket_relative_stream = None, None
+        else:
+            i_buckets_main_stream, i_bucket_relative_stream = self.cal_and_buffer_finetune_relative_positions(
+                real_positions
+            )
+        predicting_stream_pos_embed = self.embed_positions._forward(real_positions + 1)
+
+        hidden_states = self.embed_tokens(input_ids)
+        if self.embed_scale is not None:
+            hidden_states *= self.embed_scal
+        hidden_states += main_stream_pos_embed
+        # B x T x C -> T x B x C
+        hidden_states = hidden_states.transpose(0, 1)
+
+        ngram_input_embed = self.ngram_input_embed.weight
+        if use_cache:
```

Comment:
@qiweizhen @JetRunner - I think this is not fully correct here... 
IMO, it should be `if past_key_values is not None:` and not `if use_cache:`.

The problem is that by default `use_cache=True` (during inference). If someone then passes some `decoder_input_ids` with seq_length > 1, no `causal_mask` will be calculated and the result is wrong. During training `use_cache=False` anyways so there was no problem, but this should be fixed for inference IMO. 

I think we should change it to:

```python
if past_key_values not None:
       assert hidden_states.size(0) == 1, "At the moment `use_cache` is only supported for `decoder_input_ids` of length 1"
...
```

instead of 

```python
if use_cache:
...
```

Reply:
Yes it's obvious that use_cache here is not right. I agree with you.